File: solid-kata/java/dip-solution/src/main/java/com/codurance/dip/BirthdayGreeter.java
package com.codurance.dip;

import lombok.AllArgsConstructor;

import java.time.MonthDay;

@AllArgsConstructor
public class BirthdayGreeter {
    private final EmployeeRepository employeeRepository;
    private final Clock clock;
    private final EmailSender emailSender;

    public void sendGreetings() {
        MonthDay today = clock.monthDay();
        employeeRepository.findEmployeesBornOn(today)
                .stream()
                .map(this::emailFor)
                .forEach(emailSender::send);
    }

    private Email emailFor(Employee employee) {
        return Email.builder()
                .to(employee.getEmail())
                .subject("Happy birthday!")
                .message(String.format("Happy birthday, dear %s!", employee.getFirstName()))
                .build();
    }
}

File: solid-kata/c#/DIP/BirthdayGreeter.cs
namespace DIP
{
    public class BirthdayGreeter
    {
        private readonly IEmployeeRepository _employeeRepository;
        private readonly IClock _clock;

        public BirthdayGreeter(IEmployeeRepository employeeRepository, IClock clock)
        {
            _employeeRepository = employeeRepository;
            _clock = clock;
        }

        public void SendGreetings()
        {
            var today = _clock.Today();

            _employeeRepository
                .FindEmployeesBornOn(today.Month, today.Day)
                .Select(EmailFor)
                .ToList()
                .ForEach(email => new EmailSender().Send(email));
        }

        private Email EmailFor(Employee employee) =>
            new(employee.Email,
                "Happy birthday!",
                $"Happy birthday, dear {employee.FirstName}!");
    }
}